File: packages/fliwright-design-qa/android/src/main/kotlin/dev/fliwright/designqa/FliwrightDesignQaPlugin.kt
package dev.fliwright.designqa

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class FliwrightDesignQaPlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler,
    SensorEventListener {
    private lateinit var methodChannel: MethodChannel
    private lateinit var accelerometerChannel: EventChannel
    private lateinit var transportControlChannel: EventChannel
    private var sensorManager: SensorManager? = null
    private var accelerometer: Sensor? = null
    private var accelerometerSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "fliwright_design_qa/transport")
        methodChannel.setMethodCallHandler(this)

        accelerometerChannel = EventChannel(binding.binaryMessenger, "fliwright_design_qa/accelerometer")
        accelerometerChannel.setStreamHandler(this)

        transportControlChannel = EventChannel(
            binding.binaryMessenger,
            "fliwright_design_qa/transport_control",
        )
        transportControlChannel.setStreamHandler(TransportControlStreamHandler())

        sensorManager = binding.applicationContext.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        accelerometer = sensorManager?.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        result.error(
            "TRANSPORT_UNAVAILABLE",
            "The Fliwright Design QA WebRTC transport is not installed for Android.",
            "Implement the Fliwright-owned WebRTC transport before pairing a device.",
        )
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        val sensor = accelerometer
        if (sensor == null) {
            events?.error("ACCELEROMETER_UNAVAILABLE", "This device does not provide accelerometer data.", null)
            return
        }
        accelerometerSink = events
        sensorManager?.registerListener(this, sensor, SensorManager.SENSOR_DELAY_GAME)
    }

    override fun onCancel(arguments: Any?) {
        sensorManager?.unregisterListener(this)
        accelerometerSink = null
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (event.sensor.type != Sensor.TYPE_ACCELEROMETER) return
        accelerometerSink?.success(
            mapOf(
                "x" to event.values[0].toDouble(),
                "y" to event.values[1].toDouble(),
                "z" to event.values[2].toDouble(),
                "timestampMs" to System.currentTimeMillis(),
            ),
        )
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) = Unit

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        sensorManager?.unregisterListener(this)
        accelerometerSink = null
        methodChannel.setMethodCallHandler(null)
        accelerometerChannel.setStreamHandler(null)
        transportControlChannel.setStreamHandler(null)
    }
}

private class TransportControlStreamHandler : EventChannel.StreamHandler {
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) = Unit

    override fun onCancel(arguments: Any?) = Unit
}

File: packages/fliwright-design-qa/ios/Classes/FliwrightDesignQaPlugin.swift
import CoreMotion
import Flutter
import UIKit

public final class FliwrightDesignQaPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private static let accelerometerChannel = "fliwright_design_qa/accelerometer"
  private static let transportChannel = "fliwright_design_qa/transport"
  private static let transportControlChannel = "fliwright_design_qa/transport_control"
  private static let standardGravity = 9.80665

  private let motionManager = CMMotionManager()
  private var accelerometerSink: FlutterEventSink?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = FliwrightDesignQaPlugin()
    let accelerometer = FlutterEventChannel(
      name: accelerometerChannel,
      binaryMessenger: registrar.messenger()
    )
    accelerometer.setStreamHandler(instance)

    let transport = FlutterMethodChannel(
      name: transportChannel,
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: transport)

    let transportControl = FlutterEventChannel(
      name: transportControlChannel,
      binaryMessenger: registrar.messenger()
    )
    transportControl.setStreamHandler(TransportControlStreamHandler())
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    result(FlutterError(
      code: "TRANSPORT_UNAVAILABLE",
      message: "The Fliwright Design QA WebRTC transport is not installed for iOS.",
      details: "Implement the Fliwright-owned WebRTC transport before pairing a device."
    ))
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    guard motionManager.isAccelerometerAvailable else {
      return FlutterError(
        code: "ACCELEROMETER_UNAVAILABLE",
        message: "This device does not provide accelerometer data.",
        details: nil
      )
    }

    accelerometerSink = events
    motionManager.accelerometerUpdateInterval = 1.0 / 60.0
    motionManager.startAccelerometerUpdates(to: .main) { [weak self] data, error in
      guard let self else { return }
      if let error {
        self.accelerometerSink?(FlutterError(
          code: "ACCELEROMETER_ERROR",
          message: error.localizedDescription,
          details: nil
        ))
        return
      }
      guard let acceleration = data?.acceleration else { return }
      // CoreMotion reports acceleration in g; the Dart detector and Android
      // SensorManager both use m/s², so normalize iOS at the platform edge.
      self.accelerometerSink?([
        "x": acceleration.x * FliwrightDesignQaPlugin.standardGravity,
        "y": acceleration.y * FliwrightDesignQaPlugin.standardGravity,
        "z": acceleration.z * FliwrightDesignQaPlugin.standardGravity,
        "timestampMs": Int(Date().timeIntervalSince1970 * 1000),
      ])
    }
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    motionManager.stopAccelerometerUpdates()
    accelerometerSink = nil
    return nil
  }
}

private final class TransportControlStreamHandler: NSObject, FlutterStreamHandler {
  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    return nil
  }
}
